pkg/middleware/instrumentation.go

```diff
@@ -0,0 +1,109 @@
+package middleware
+
+import (
+	"net/http"
+
+	"github.com/justinas/alice"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+)
+
+var metricsHandler = NewInstrumentMetricHandlerWithDefaultRegistry()
+
+func NewInstrumentMetricHandlerWithDefaultRegistry() http.Handler {
+	return NewInstrumentMetricHandler(prometheus.DefaultRegisterer, prometheus.DefaultGatherer)
+}
+
+func NewInstrumentMetricHandler(registerer prometheus.Registerer, gatherer prometheus.Gatherer) http.Handler {
+	return promhttp.InstrumentMetricHandler(
+		registerer, promhttp.HandlerFor(gatherer, promhttp.HandlerOpts{}),
+	)
+}
+
+func NewRequestInstrumentorWithDefaultRegistry() alice.Constructor {
+	return NewRequestInstrumentor(prometheus.DefaultRegisterer)
+}
+
+func NewRequestInstrumentor(registerer prometheus.Registerer) alice.Constructor {
+	return func(next http.Handler) http.Handler {
+		// Counter for all requests
+		// This is bucketed based on the response code we set
+		counterHandler := func(next http.Handler) http.Handler {
+			return promhttp.InstrumentHandlerCounter(registerRequestsCounter(registerer), next)
+		}
+
+		// Gauge to all requests currently being handled
+		inFlightHandler := func(next http.Handler) http.Handler {
+			return promhttp.InstrumentHandlerInFlight(registerInflightRequestsGauge(registerer), next)
+		}
+
+		// The latency of all requests bucketed by HTTP method
+		durationHandler := func(next http.Handler) http.Handler {
+			return promhttp.InstrumentHandlerDuration(registerRequestsLatencyHistogram(registerer), next)
+		}
+
+		return alice.New(counterHandler, inFlightHandler, durationHandler).Then(next)
+	}
+}
+
+func registerRequestsCounter(registerer prometheus.Registerer) *prometheus.CounterVec {
+	counter := prometheus.NewCounterVec(
+		prometheus.CounterOpts{
+			Name: "oauth2_proxy_requests_total",
+			Help: "Total number of requests by HTTP status code.",
+		},
+		[]string{"code"},
+	)
+
+	if err := registerer.Register(counter); err != nil {
+		if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
+			counter = are.ExistingCollector.(*prometheus.CounterVec)
+		} else {
+			panic(err)
+		}
+	}
+
+	return counter
+}
+
+func registerInflightRequestsGauge(registerer prometheus.Registerer) prometheus.Gauge {
+	gauge := prometheus.NewGauge(prometheus.GaugeOpts{
+		Name: "oauth2_proxy_requests_in_flight",
+		Help: "Current number of requests being served.",
+	})
+
+	if err := registerer.Register(gauge); err != nil {
+		if are, ok := err.(prometheus.AlreadyRegisteredError); ok {
+			gauge = are.ExistingCollector.(prometheus.Gauge)
+		} else {
+			panic(err)
```

Comment:
Is `panic` correct in these scenarios? Or should we control the `logger.Errorf` or `logger.Fatalf` in a more structured manner before we either panic or exit?

Reply:
Oh nevermind, I think I see the documentation on how prometheus behaves if the wrapped handler panics.